crates/key-derivation/src/sequencer.rs

```diff
@@ -0,0 +1,200 @@
+//! Key derivation for Strata sequencer
+
+use bitcoin::bip32::{ChildNumber, DerivationPath, Xpriv, Xpub};
+use secp256k1::SECP256K1;
+#[cfg(feature = "zeroize")]
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+use crate::error::KeyError;
+
+/// The Strata base index for sequencer keys.
+const BASE_IDX: u32 = 56;
+
+/// The Strata sequencer index.
+const SEQUENCER_IDX: u32 = 10;
+
+/// The Strata sequencer's master, and derived _private_ keys.
+#[derive(Debug, Clone)]
+pub struct SequencerKeys {
+    /// Sequencer's master [`Xpriv`].
+    master: Xpriv,
+
+    /// Sequencer's derived [`Xpriv`].
+    derived: Xpriv,
+}
+
+impl SequencerKeys {
+    /// Creates a new [`SequencerKeys`] from a master [`Xpriv`].
+    pub fn new(master: &Xpriv) -> Result<Self, KeyError> {
+        let path = derived_path();
+
+        let derived = master.derive_priv(SECP256K1, &path?)?;
+        Ok(Self {
+            master: *master,
+            derived,
+        })
+    }
+
+    /// Sequencer's master [`Xpriv`].
+    pub fn master_xpriv(&self) -> &Xpriv {
+        &self.master
+    }
+
+    /// Sequencer's derived [`Xpriv`].
+    pub fn derived_xpriv(&self) -> &Xpriv {
+        &self.derived
+    }
+
+    /// Sequencer's master [`Xpub`].
+    ///
+    /// Infallible conversion from [`Xpriv`] to [`Xpub`] according to
+    /// [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+    pub fn master_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.master)
+    }
+
+    /// Sequencer's derived [`Xpub`].
+    ///
+    /// Infallible conversion from [`Xpriv`] to [`Xpub`] according to
+    /// [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+    pub fn derived_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.derived)
+    }
+}
+
+// Manual Drop implementation to zeroize keys on drop.
+impl Drop for SequencerKeys {
+    fn drop(&mut self) {
+        #[cfg(feature = "zeroize")]
+        self.zeroize();
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl Zeroize for SequencerKeys {
+    #[inline]
+    fn zeroize(&mut self) {
+        let Self { master, derived } = self;
+
+        // # Security note
+        //
+        // Going over all possible "zeroizable" fields.
+        // What we cannot zeroize is only:
+        //
+        // - Network: enum
+        //
+        // These are fine to leave as they are since they are public parameters,
+        // and not secret values.
+        //
+        // NOTE: `Xpriv.private_key` (`SecretKey`) `non_secure_erase` writes `1`s to the memory.
+
+        // Zeroize master components
+        master.depth.zeroize();
+        {
+            let fingerprint: &mut [u8; 4] = master.parent_fingerprint.as_mut();
+            fingerprint.copy_from_slice(&[0u8; 4]);
```

Comment:
See comment above about mutable slice zeroizing.

Reply:
Done in 55a0288c37ab075ee12b5c910d367fafd2c2cbfa